### ISSUE TITLE
[Chains] refactor eris chains to use an interface

### ISSUE THREAD

**Issue by compleatang** [monax/cli#336]
which should be pluggable by diff chain types. relates to #200 .

**Comment by compleatang** [comment]
@pietv after the windows fixes, if you don't have anything higher priority this is creeping up on my priority list. as OBC starts to come online, as well as other chain types it will be important for us to ready to support them.

**Comment by pietv** [comment]
@compleatang gotcha

**Comment by compleatang** [comment]
its not urgent or anything. i know you also have the services/perform tests separation on your plate as well...

**Comment by compleatang** [comment]
@pietv if we should bump this to 0.12 go ahead.

**Comment by pietv** [comment]
This is an outline of how it would look like after moving to use a chain interface instead
of a single chain definition struct. The reasons for that move are the following:

- The necessity to use at least 2 different chain implementations
(engines) and be able to specify which one to use 
(`eris chains make --type geth`) and be able to quickly add others.
- The need to declutter chain management code of setting environment variables, importing and exporting directories, boot sequence logic, presence of file names and paths that
might be specific only to one chain implementation
and not the other (https://github.com/eris-ltd/eris-cli/blob/master/chains/operate.go#L227-L412).
- Improve testability (by making functions simpler) and test coverage of chain management code (the use of
an interface allows for an easy mocking).
- Hide chain dependency on Docker container running code
(the next chain implementation is going to be run in a Docker container still, but doesn't have to be).

The proposed solution is:

- To use a chain interface with method names similar to the current
`eris chains` commands with less similarity to Docker operations and removed repetition (more [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design))).
- The chain implementation dependent actions (boot sequence,
imports, etc.) reside in
implementation dependent configuration files as text
(the sample names for which are `engines/db.toml` and `engines/geth.toml` files)
not in code.
- Extract `do` definition values which are chain implementation
dependent (such as `GenesisFile`, `DefaultGas`, `AccountTypes`)
into multiple engine dependent structs.
- Remove the currently used chain definition altogether:
the implementation dependent configuration file (engine file)
will be loaded in an engine struct which will retain enough
information to interact with `perform` package actions (which
uses `service` and `operations` structs to pass parameters).


## Chain interface

Chain interface defines operations on _some thing_ that can
do chain related operations.

Four operations are not part of the interface and are intended
to either set up a new chain or load an existing one:

```
type chain interface{}

chains.Make(name, type) error
chains.New(name, type) (error, chain)
chains.Import(name) (error, chain)
chains.Load(name) (error, chain)
```

`Make` calls the `cm` package to generate `type` chain
configuration files on host.

`New` calls the `cm` package to parse the files into a struct and
creates the chain (runs Docker containers,
imports data).

`Import` corresponds to the `chains import` command;
detects the type of the chain; puts the chain files locally then calls `New` to create the chain.

`Load` calls the `cm` package to make sense of the files
in the chain directory (parses the files into a struct ); doesn't do anything else.

The other operations are interface methods which roughly
correspond to `eris chains` commands:


- `Start(interface{}) error` - start the chain (`eris chains start`)
- `Stop() error` - stop the chain (`eris chains stop`)
- `Logs(...) (error, *io.Reader)` - retrieve chain logs (`eris chains logs`). _Likely can be merged with the next command_.
- `Commands(...) (error, *io.Reader)` - run chain dependent commands (`eris chains cat`, `eris chains inspect`, `eris chains ports`).

The following `eris chains` commands are implemented using the already defined

- `Stop(); Start()` - `eris chains restart`
- `Remove(); New()` - `eris chains update`
- `Start(...)` - `eris chains exec`


(The not mentioned `chains export` command doesn't need to be implemented using the chain interface.)

## Engine configuration file

The chain implementation dependent configuration file
sketch resembles the chain definition file (or the `config.toml` file) with a few additions:

```
[service]
image          = "quay.io/eris/db:{{.Version}}"
ports          = [ "1337", "46656", "46657" ]
entry_point    = "erisdb-wrapper"
environment    = [
                   "ERISDB_API=true",
                   "CONTAINER_NAME={{.ContainerName}}",
                   "CHAIN_ID={{.ChainID}}",
                   "CONFIG_OPTS={{.Options}}",
                   "NODE_ADDR={{.Gateway}}"
                 ]

[dependencies]
data     = true
services = [ "keys" ]

[[imports]]
dependency = "data"
script = `mkdir -p {{.ChainPath}}`

[[imports]]
dependency = "data"
file = [ "genesis.json", "{{.ChainPath}}" ]

[[imports]]
dependency = "data"
file = [ "priv_validator.json", "{{.ChainPath}}"]

[[imports]]
dependency = "data"
file = [ "config.toml", "{{.ChainPath}}"]

[[imports]]
dependency = "keys"
script = `mintkey eris {{.ChainPath}}/priv_validator.json`

[[imports]]
dependency = "keys"
script = `chown --recursive eris {{.ErisRoot}}`

[[commands]]

  [cat.genesis]
  description = "display chain's genesis.json file"
  shell = `cat {{.ChainPath}}/genesis.json`

  [cat.config]
  description = "display chain's config.toml configuration file"
  script = `cat {{.ChainPath}}/config.toml`

  [cat.status]
  description = "display chain status"
  script = `mintinfo --node-addr http://chain:46657 status`

  [cat.validators]
  description = "display chain validators"
  script = `mintinfo --node-addr http://chain:46657 validators`

  [cat.test]
  description = "test command"
  script = `echo {{index .Args 0}} {{index .Args 1}}`

[maintainer]
name = "Eris Industries"
email = "support@erisindustries.com"
```

When `chains.New(type)` operation populates the chain engine struct
before creating the chain (starting Docker containers):

```
type DBEngine struct {
        Engine engine
        Chain  chain
}
```

The `Engine` struct is roughly the old chain definition file +
engine dependent fields extracted from the `Do` definition +
standard paths and directories.

The `Chain` struct is what the `cm` package was able to parse
while reading the generated text files on the host.
(The structure may be directly imported from the `cm`, depending
on its implementation.)

Both these structs are exposed via Go templates to be used
as fields in the engine configuration file.

A few other clarifications:

- The `[[name]]` TOML syntax is an array of maps. The `[[imports]]` are
arranged as an array to be able to execute import operations in sequence
and be flexible in which order which container to import to
(import to `data` then to `keys` then again to `data`, etc.).

- The `data = true` is the suggestion to refactor of
the `data_container=true` because the latter is a dependency.

- For the `[[imports]]` clause, when the dependency
is a data container (the assumption is that only one
data container to be used), the `dependency = "data"`
is omitted (to be the default).

- The `[[commands]]` section are commands specific
to this chain implementation only to be run
on the data container of the running chain and to
be exposed as `eris chains <ENGINE_NAME> <CHAIN_NAME> cat validators`,
for example. The presence of the `ENGINE_NAME` seems necessary
(suggestions welcome!) to be able to build the Cobra
commands _before_ the chain type is detected or
issued (`eris --help` should be able to show them).

- The scripts in `[[imports]]` and `[[commands]]` don't
have to be shell scripts. The functionality could be
implemented using the powerful Go template helper functions
by manipulating the exposed struct fields
(tentatively named `Engine` and `Chain`).
Later they cuold also benefit from using `actions` (`eris actions`).

## Chain type detection

The `chains.Load()` and `chains.Import()` operations should
be able to detect the chain type without asking for `cm`
package's help. Having a special configuration file for this
single bit of information seems unnecessary.

One alternative is to use an additional
container label (for example, **CHAIN_TYPE**), which makes
it easy to use with `ls` functions,
but unfortunately the labels is a Docker feature and
the goal is to make the chains agnostic of that; it also makes for simpler stub mock chain implementations.

A second alternative is to use a special prefix in chain
directory names as in `name.type` (`simplechain.erisdb`)
which subjectively seems ugly.

Another alternative is to use a nested directory structure
`type/name` (`erisdb/simplechain`) the downside of which
is having to traverse the directories (deeper manual walks).

Other than having to detect a chain type, there is no need for a chain configuration file.

## Chains make refactoring

Currently the `chains make` command is implemented using
a separate Docker container; it is going to be refactored
into a package (https://github.com/eris-ltd/eris-cli/issues/899).

One can think of one requirement for the refactored `cm` package
that _generate files_ and _parse files_ be separate independent
operations (`cm.Make()` and `cm.Load()` something).

The parse operation is needed to be able to implement
`[[commands]]` on a host rather
than a container. For example, this ticket https://github.com/eris-ltd/eris-cli/issues/759
would be easily implemented with something like

```
[[commands]]
  [info.address]
  host = `{{.Chain.Address}}`
```

**Comment by pietv** [comment]
@zramsay @compleatang @benjaminbollen please share your thoughts.

**Comment by zramsay** [comment]
Some notes:
- Since we already have `--chain-type` and it probably shouldn't change, I'd suggest `eris chains make --kind geth/edb`. No strong opinions here though. `--engine` is good too
- the `func setupChain` has been somewhat cleaned up already in a feature branch (#feature-sort-configs-and-loaders) using the new [cm config](https://github.com/eris-ltd/eris-cm/pull/31/files), alongside relevant refactor of chains code for #651 
- Do we need engine files? Perhaps ecm can hold the implementation details in the `config.toml` it writes.
- ++ on removing the chain definition files altogether. This is mostly done in the above mentioned branch, where the chain def and `default.toml` have been removed in favour of ecm's single `config.toml`.
- Note that `chains import/export` have been deprecated for the time being.
- WRT to the engine config file, we need to sync with @benjaminbollen. Either ecm should write this on `chains make` or cli updates the file itself before starting the chain.
- Not convinced we need to `eris chains <ENGINE_NAME> <CHAIN_NAME>` since the engine can be detected either via 1) reading from the config.toml and/or 2) using `CHAIN_TYPE` (preferable renamed to avoid confusion with `--chain-type`) as a docker label. FTR, I am in favour of the latter (well, both) for preserving the functionality with `ls`. nested/prefixed directories would be ugly IMO.
- down the road we're going to want `--consensus-type=tendermint/casper/etc` so that placeholder should perhaps be included in the interface.

**Comment by pietv** [comment]
> Since we already have --chain-type and it probably shouldn't change, I'd suggest eris chains make --kind geth/edb. No strong opinions here though. --engine is good too
++

> the func setupChain has been somewhat cleaned up already in a feature branch
very nice! this change is about cleaning this further by moving setting environment, boot dependencies, import of files to a configuration file altogether.

> Do we need engine files? Perhaps ecm can hold the implementation details in the config.toml it writes.
We need variable configuration files depending on which engine is selected. In my view, low level information - container logic, custom commands and order of booting containers is the information to hold on the `cli` side. "Business logic" - number of nodes, genesis and validator file contents better be hold on the `cm` side. Besides, the purpose of the engine config is to implement informational commands like (`chains inspect`) which should be available for displaying help when there are no chains created yet.

> Note that chains import/export have been deprecated for the time being.
++

> Not convinced we need to `eris chains <ENGINE_NAME> <CHAIN_NAME>` since the engine can be detected either via 1) reading from the config.toml and/or 2) as a docker label.
I'd like the `eris chains <ENGINE_NAME ...` commands be available when no chain is created, so no `config.toml` or Docker labels.
Agree with renaming the `CHAIN_NAME`.

> down the road we're going to want --consensus-type=tendermint/casper/etc so that placeholder should perhaps be included in the interface.
Good to have this setting is the files generated by `cm.Make`, agree.

**Comment by compleatang** [comment]
My thoughts:

## Chain Interface

* `chains.Make(name, type) error` -> `chains.Make(name, type) (error, chain)` (this will give us better programmatic access when we start on agents)

> New calls the cm package to parse the files into a struct and creates the chain ...

^^ excellent. let's move all that chain management stuff down to the `cm` library.

> ... (runs Docker containers, imports data).

Hmm. Not sure that `cm` level should be working with Docker containers and imports. What's the argument/thought here?

* `Logs(...) ... Likely can be merged with the next.` <- yes. do merge `Logs` into `Commands`. this will give `eris chains` flexibility in where and how they do their logging.

Rest is A-OK!

## Engine config file

```toml
# moving data to dependencies is great!
[dependencies]
data     = true
```

```toml
# imports.file is super. this is very, very clean. we can also use this for services.
[[imports]]
dependency = "data"
file = [ "config.toml", "{{.ChainPath}}"]

# imports.script is also super!
# the use of gotemplates here is powerful and great! 
[[imports]]
dependency = "keys"
script = `mintkey eris {{.ChainPath}}/priv_validator.json`
```

Those two `imports.file` and `imports.script` are very clean. This is a nice framework that we can optimize cli around. 

```toml
# clean. super!
[[commands]]

  [cat.genesis]
  description = "display chain's genesis.json file"
  shell = `cat {{.ChainPath}}/genesis.json`
```

> The Engine struct is roughly the old chain definition file + engine dependent fields extracted from the Do definition + standard paths and directories.

I'm A-OK for us starting to break up that do struct where it makes sense to. 

> The Chain struct is what the cm package was able to parse while reading the generated text files on the host. (The structure may be directly imported from the cm, depending
on its implementation.)

Great. And we should be working to minimize those generated text files. We should be getting it down to genesis.json && config.toml very shortly. @zramsay and I talked about this and @benjaminbollen and I *will be* ... :-) 

* At first I was resistant to this idea of an engine config file at all if I'm honest. We've been trying to reduce our configuration files. However, I like the ideas here. I think these should live in `common.ChainTypePath` (and I also think the existing chainTypes can perhaps use a rethinking at some point).
* What I think we should do is to read the engine config file first as our defaults, but allow the config.toml to overwrite. I'm pretty sure the structure you're proposing does that, but maybe I'm wrong. 

> The [[commands]] section are commands specific to this chain implementation only to be run on the data container of the running chain ...

Agree fully. Like it.

> and to be exposed as eris chains <ENGINE_NAME> <CHAIN_NAME> cat validators, for example. The presence of the ENGINE_NAME seems necessary (suggestions welcome!) ...

I'm not totally following the logic here. Wouldn't it be better to cache a chain_name -> engine_name mapping somewhere? We may also want to log other things (`chain_id` for example) that make for easy reference by the components of the stack. 

* In general I think this is clean. I like it a lot. One thing you are going to want to add into your thinking is that at some point we're going to need the following levels of configuration:
  * `engine` (this is a nice setup and I'm happy with this)
  * `chain` (ben's config file system is terrific now)
  * `network` (Zach and Rick will hopefully be working on this soon-ish) <- should load port settings, DNS, most of the docker ports settings should be exposed and manipulated via those config files

## Chain type detection

See above for my thoughts on this. I mostly agree with this section.

(From your reply to Zach)...

> We need variable configuration files depending on which engine is selected. In my view, low level information - container logic, custom commands and order of booting containers is the information to hold on the cli side. "Business logic" - number of nodes, genesis and validator file contents better be hold on the cm side.

For me this is the cleanest point of separation as well. 

> My suggestion is also to move away from storing information in config.toml that is not directly engine related (like container image - the engine would be able to run as a binary outside Docker).

Yes!

## Others

From Zach:

> Note that chains import/export have been deprecated for the time being.

Zach's talking about the importing of files from IPFS. Your `chains.Import(...)` is different obvi.

From You:

> I'd like the eris chains <ENGINE_NAME ... commands be available when no chain is created, so no config.toml or Docker labels.

If we just keep a csv file cached in `common.ChainsPath` that with a mapping as suggested above we can still allow for those commands without a config or Docker labels. :-)

**Comment by compleatang** [comment]
I've also converted this to an Epic so we can add smaller issues to it.

**Comment by VoR0220** [comment]
Awesome

**Comment by compleatang** [comment]
Alas....
